### PR TITLE
Make libcloud compliant with the latest Outscale API.

### PR DESCRIPTION
**Pull request by outscale-fne**
### Description

Added the API Acess Rules and Client Certificate Authority (CA) enpoints.

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation

### CODE REVIEW

**Comment by outscale-fne** [comment]
Up!

File: libcloud/compute/drivers/outscale.py [hunk 1]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
             return response.json()["VpnConnections"]
         return response.json()
 
+    def ex_create_certificate_authority(
+        self,
+        ca_perm: str = None,
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: If the argument is required, it should not default to None.

File: libcloud/compute/drivers/outscale.py [hunk 2]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        :param      description: The description of the CA.
+        :type       description: ``bool``
+
+        :param      dry_run: If true, checks whether you have the required
+        permissions to perform the action.
+        :type       dry_run: ``bool``
+        """
+        action = "CreateCa"
+        data = {"DryRun": dry_run}
+        if ca_perm is not None:
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: Hm, I'm a bit confused, docstring says it's required, but code makes it look like it's not.

File: libcloud/compute/drivers/outscale.py [hunk 3]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        if description is not None:
+            data.update({"Description": description})
+        response = self._call_api(action, json.dumps(data))
+        if response.status_code == 200:
+            return response.json()["Ca"]
+        return response.json()
+
+    def ex_delete_certificate_authority(
+        self,
+        ca_id: str = None,
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: I assume id is required so it should not default to None.

File: libcloud/compute/drivers/outscale.py [hunk 4]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        :param      ca_id: The ID of the CA you want to delete. (required)
+        :type       ca_id: ``str``
+
+        :param      dry_run: If true, checks whether you have the required
+        permissions to perform the action.
+        :type       dry_run: ``bool``
+        """
+        action = "DeleteCa"
+        data = {"DryRun": dry_run}
+        if ca_id is not None:
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: If ca_id is required, this if check should not be there.

File: libcloud/compute/drivers/outscale.py [hunk 5]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        ca_fingerprints: List[str] = None,
+        ca_ids: List[str] = None,
+        descriptions: List[str] = None,
+        dry_run: bool = False
+    ):
+        """
+        Returns information about one or more of your Client Certificate
+        Authorities (CAs).
+
+        :param      ca_fingerprints: The fingerprints of the CAs.
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py:7776]: Which arguments in this method are required? I assume either ca_ids or ca_fingersprints? And perhaps they are even mutually exclusive? Or what happens if both are provided? Same with descriptions.

File: libcloud/compute/drivers/outscale.py [hunk 6]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        :param      descriptions: The descriptions of the CAs.
+        :type       descriptions: ``list`` of ``str``
+
+        :param      dry_run: If true, checks whether you have the required
+        permissions to perform the action.
+        :type       dry_run: ``bool``
+        """
+        action = "ReadCas"
+        data = {"DryRun": dry_run, "Filters": {}}
+        if ca_fingerprints is not None:
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py:7794]: I assume at least one of those filters needs to be provided, right? If so, we should check that in the code and throw if it's not true.

File: libcloud/compute/drivers/outscale.py [hunk 7]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        if descriptions is not None:
+            data["Filters"].update({"Descriptions": descriptions})
+        response = self._call_api(action, json.dumps(data))
+        if response.status_code == 200:
+            return response.json()["Cas"]
+        return response.json()
+
+    def ex_update_certificate_authority(
+        self,
+        ca_id: str = None,
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: Same here, required argument should not default to None.

File: libcloud/compute/drivers/outscale.py [hunk 8]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        description: str = None,
+        ip_ranges: List[str] = None,
+        ca_ids: List[str] = None,
+        cns: List[str] = None,
+        dry_run: bool = False,
+    ):
+        """
+        Create an API access rule.
+        It is a rule to allow access to the API from your account.
+        You need to specify at least the CaIds or the IpRanges parameter.
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py:7848]: Please remove default None value for required arguments.

File: libcloud/compute/drivers/outscale.py [hunk 9]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+        if cns is not None:
+            data["Cns"] = cns
+        response = self._call_api(action, json.dumps(data))
+        if response.status_code == 200:
+            return response.json()["ApiAccessRule"]
+        return response.json()
+
+    def ex_delete_api_access_rule(
+        self,
+        api_access_rule_id: str = None,
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: Same here, None should be removed.

File: libcloud/compute/drivers/outscale.py [hunk 10]
```diff
@@ -7709,6 +7709,318 @@ def ex_list_vpn_connections(
...
+            filters["IpRanges"] = ip_ranges
+        data = {"Filters": filters, "DryRun": dry_run}
+        response = self._call_api(action, json.dumps(data))
+        if response.status_code == 200:
+            return response.json()["ApiAccessRules"]
+        return response.json()
+
+    def ex_update_api_access_rule(
+        self,
+        api_access_rule_id: str = None,
```
Comment (Kami) [inline libcloud/compute/drivers/outscale.py]: Please remove None from a required argument.

**Review by Kami — CHANGES_REQUESTED** [review]
Thanks for the contribution.

I've added some in-line comments.

Mostly needs some argument default value clean up and adding some additional checks to the method where at least one filter argument needs to be provided.

**Comment by outscale-fne** [comment]
> Thanks for the contribution.
> 
> I've added some in-line comments.
> 
> Mostly needs some argument default value clean up and adding some additional checks to the method where at least one filter argument needs to be provided.

Thank you sir for reviewing! It seems we misunderstood a part af the conding rules... I am patching right now!

File: libcloud/compute/drivers/outscale.py (hunk 5 above)
Comment (outscale-fne) [reply libcloud/compute/drivers/outscale.py:7776]: Here all arguments are not required. You would just get all the certificate authorities you have. If you use exery args, you will get all the certificate authorities matching your requirements. All these params are filters.

File: libcloud/compute/drivers/outscale.py (hunk 6 above)
Comment (outscale-fne) [reply libcloud/compute/drivers/outscale.py:7794]: As said before, no args required here.

File: libcloud/compute/drivers/outscale.py (hunk 8 above)
Comment (outscale-fne) [reply libcloud/compute/drivers/outscale.py:7848]: `You need to specify at least the CaIds or the IpRanges parameter.`
I don't knbow what to do here.
If I remove the default None for both, the user will have to set one to none himself right? Is that what you are asking me to do?
To be sure I understood everything!

File: libcloud/compute/drivers/outscale.py (hunk 6 above)
Comment (Kami) [reply libcloud/compute/drivers/outscale.py:7794]: Thanks for clarifying - I thought at least one filter must be provided.

File: libcloud/compute/drivers/outscale.py (hunk 8 above)
Comment (Kami) [reply libcloud/compute/drivers/outscale.py:7848]: Yeah, in this case you should default them to None and then you can add check similar to this one to the code:

```python
if not ca_ids and not ip_ranges:
    raise ValueError("Either ca_ids or ip_ranges argument must be provided")
```

Also, are those arguments mutually exclusive? If they are, we should also add another check:

```python
if ca_ids and ip_ranges:
    raise ValueError("Either ca_ids or ip_ranges argument must be provided, but not both")
```

**Comment by Kami** [comment]
Thanks for addressing the feedback - there is one small last comment which still need some work.

File: libcloud/compute/drivers/outscale.py (hunk 8 above)
Comment (outscale-fne) [reply libcloud/compute/drivers/outscale.py:7848]: Thank you for the explanations. I am addressing this right now.

**Review by outscale-fne — COMMENTED** [review]
Seems ok now.

**Review by Kami — APPROVED** [review]

**Comment by Kami** [comment]
Thanks. Merged into trunk.

In the future, please also make sure you add corresponding unit tests (including for all the various edge cases - e.g. ensuring ValueError is thrown, etc.).
